File: Problems/200. Number of Islands/number-of-islands.java
//Solution 1: 4ms 49.48 MB
class Solution {
  private static final int[][] DIRECTIONS = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

  private void clearIsland(int row, int col, char[][] grid) {
    if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length
        || grid[row][col] == '0') {
      return;
    }

    grid[row][col] = '0';

    for (int[] direction : DIRECTIONS) {
      clearIsland(row + direction[0], col + direction[1], grid);
    }
  }

  public int numIslands(char[][] grid) {
    int islandCount = 0;

    for (int i = 0; i < grid.length; i++) {
      for (int j = 0; j < grid[0].length; j++) {
        if (grid[i][j] == '1') {
          clearIsland(i, j, grid);
          islandCount++;
        }
      }
    }

    return islandCount;
  }
}

//Solution 2: 3ms 49.02 MB
class Solution {
  public void clearIsland(char[][] grid, int row, int col) {
    if (row < 0 || row == grid.length || col < 0 || col == grid[0].length
        || grid[row][col] == '0')
      return;

    grid[row][col] = '0';
    clearIsland(grid, row - 1, col); // top
    clearIsland(grid, row, col - 1); // left
    clearIsland(grid, row + 1, col); // bottom
    clearIsland(grid, row, col + 1); // right

    return;
  }

  public int numIslands(char[][] grid) {
    int islandCount = 0;

    for (int row = 0; row < grid.length; row++) {
      for (int col = 0; col < grid[0].length; col++) {
        if (grid[row][col] == '1') {
          clearIsland(grid, row, col);
          islandCount++;
        }
      }
    }

    return islandCount;
  }
}

//Solution 3: 2ms 49.38 MB
class Solution {
  int nr = 0;
  int nc = 0;

  public void clearIsland(char[][] grid, int row, int col) {
    if (row < 0 || row == nr || col < 0 || col == nc || grid[row][col] == '0')
      return;

    grid[row][col] = '0';
    clearIsland(grid, row - 1, col); // top
    clearIsland(grid, row, col - 1); // left
    clearIsland(grid, row + 1, col); // bottom
    clearIsland(grid, row, col + 1); // right

    return;
  }

  public int numIslands(char[][] grid) {
    int islandCount = 0;
    nr = grid.length;
    nc = grid[0].length;

    for (int row = 0; row < nr; row++) {
      for (int col = 0; col < nc; col++) {
        if (grid[row][col] == '1') {
          clearIsland(grid, row, col);
          islandCount++;
        }
      }
    }

    return islandCount;
  }
}

File: Problems/200. Number of Islands/number-of-islands.ts
//Solution 1: 71ms 53.14 MB
const directions = [
  [-1, 0],
  [1, 0],
  [0, -1],
  [0, 1],
];

function clearIsland(x: number, y: number, grid: string[][]): void {
  const xSize = grid.length;
  const ySize = grid[0].length;

  if (x >= xSize || x < 0 || y >= ySize || y < 0 || grid[x][y] === "0") {
    return;
  }

  grid[x][y] = "0"; // Mark current cell as visited

  for (const direction of directions) {
    clearIsland(x + direction[0], y + direction[1], grid);
  }
}

function numIslands(grid: string[][]): number {
  const xSize = grid.length;
  const ySize = grid[0].length;

  let islands = 0;

  for (let i = 0; i < xSize; i++) {
    for (let j = 0; j < ySize; j++) {
      if (grid[i][j] === "1") {
        clearIsland(i, j, grid);
        islands++;
      }
    }
  }

  return islands;
}

//Solution 2: 65ms 53.46 MB
function numIslands(grid: string[][]): number {
  const xSize = grid.length;
  const ySize = grid[0].length;
  const directions = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  let islands = 0;

  function clearIsland(x: number, y: number): void {
    if (x < 0 || x >= xSize || y < 0 || y >= ySize || grid[x][y] === "0") {
      return;
    }

    grid[x][y] = "0";

    for (const direction of directions) {
      clearIsland(x + direction[0], y + direction[1]);
    }
  }

  for (let i = 0; i < xSize; i++) {
    for (let j = 0; j < ySize; j++) {
      if (grid[i][j] === "1") {
        clearIsland(i, j);
        islands++;
      }
    }
  }

  return islands;
}

//Solution 3: 55ms 52.88 MB
function numIslands(grid: string[][]): number {
  const xSize = grid.length;
  const ySize = grid[0].length;

  let islands = 0;

  function clearIsland(x: number, y: number): void {
    if (x < 0 || x >= xSize || y < 0 || y >= ySize || grid[x][y] === "0") {
      return;
    }

    grid[x][y] = "0";

    clearIsland(x + 1, y);
    clearIsland(x - 1, y);
    clearIsland(x, y + 1);
    clearIsland(x, y - 1);
  }

  for (let i = 0; i < xSize; i++) {
    for (let j = 0; j < ySize; j++) {
      if (grid[i][j] === "1") {
        clearIsland(i, j);
        islands++;
      }
    }
  }

  return islands;
}
